plutus-benchmark/flat/Codec.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PackageImports    #-}
+
+module Codec ( Codec(..)
+             , flatCodec
+             , cborCodec
+             , withZlib
+             , withPureZlib
+             , codecs
+             ) where
+
+import           Control.DeepSeq                     (NFData, force)
+import qualified Data.ByteString                     as BS
+import qualified Data.ByteString.Lazy                as LBS
+import           Data.Either                         (fromRight)
+import           Data.Text                           (Text)
+
+import           Language.PlutusCore                 (Name (..), Unique (..))
+import           Language.PlutusCore.Universe
+import           Language.UntypedPlutusCore
+import           Language.UntypedPlutusCore.DeBruijn
+
+import           "pure-zlib" Codec.Compression.Zlib  as PureZlib
+import           "zlib" Codec.Compression.Zlib       as Zlib
+import           Codec.Serialise                     (deserialise, serialise)
+import           Flat                                (DecodeException (..), flat, unflat)
+
+type Tm = Term Name DefaultUni ()
+
+data Codec a = Codec
+  { serialize   :: a -> BS.ByteString
+  , deserialize :: BS.ByteString -> a
+  }
+
+fromDecoded :: Show error => Either error a -> a
+fromDecoded (Left err) = error $ show err
+fromDecoded (Right  v) = v
+
+flatCodec :: Codec (Tm)
+flatCodec = Codec
+  { serialize   = force . flat
+  , deserialize = force . fromDecoded . unflat
+  }
+
+cborCodec :: Codec (Tm)
+cborCodec = Codec
+  { serialize   = force . LBS.toStrict . serialise
+  , deserialize = force . deserialise . LBS.fromStrict
+  }
+
+withZlib :: NFData a => Codec a -> Codec a
+withZlib codec = Codec
+  { serialize = force . LBS.toStrict . Zlib.compress . LBS.fromStrict . serialize codec
```

Comment:
Note that you can change the compression level here: see https://hackage.haskell.org/package/zlib-0.6.2.2/docs/Codec-Compression-Zlib.html#g:2

I quickly tried the best compression level.  It increased the serialisation time (in a few cases by a factor or three or four), but didn't make any apparent difference to the decompression time.  This is what you'd expect: when you're compressing you have to analyse the global structure of the input, but when you're decompressing you just process the input as a stream.  It didn't make a great deal of difference to the compressed sizes compared to the default compression method though.  It just saved a few bytes in each case, the biggest saving being for the future contract, where the size of the compressed flat encoding decreased from 16133 bytes to 16102, saving a massive 31 bytes.

Reply:
From what you are saying there are no major savings on size, and we take a hit when zipping. I do not see why I would want to change the compression level...

Reply:
Note that we don't care about serialization time, only deserialization.

Reply:
> From what you are saying there are no major savings on size, and we take a performance hit when zipping. I do not see why I would want to change the compression level.

Exactly!  I thought it was worth trying, but it turned out that the gains were insignificant.